enhancements/ingress/set-delete-http-headers.md

```diff
@@ -0,0 +1,1120 @@
+| `x-ssl`                    | None                                                | Yes                                                               | Yes                                                   | None                                                                        | None                                                 |                    
+
+
+### Drawbacks
+
+Please refer to the `Risks and Mitigations` section to understand which headers are not allowed to be set/delete using this EP.
+
+## Design Details
+
+### Open Questions
+
+#### Will this enhancement cover setting or deleting request headers in addition to response headers?
+  Answer: Yes, we will be covering setting or deleting of request headers.
+  This is required for the following use-case.  
+  
+  As a cluster admin or route owner, I want to enable mutual TLS but replace the "x-ssl-client-der" request header
+  that OpenShift router adds with an "x-forwarded-client-cert" request header that my application checks for, 
+  so that I can migrate my applications that rely on mutual TLS and use the latter header name. 
+
+  The following example replaces "x-ssl-client-der" with "x-forwarded-client-cert":
+```yaml
+spec:
+  httpHeaders:
+    actions:
+      response:
+      - name: X-SSL-Client-Cert
+        action:
+        type: Set
+        set:
+          value: "%{+Q}[ssl_c_der,base64]"
+      - name:   Content-Language
+        action:
+        type: Delete
+```
+
+#### Will this enhancement cover setting the load-balancing algorithm based on headers?
+  Answer: The HAProxy configuration provides a [load balancing algorithm](http://cbonte.github.io/haproxy-dconv/2.6/configuration.html#4.2-balance) which relies on the
+  check against the incoming request's header values.
+  
+  Even though this enhancement adds the header manipulation capability it
+  doesn't aim at providing (or enhancing) the load balancing API for the IngressController or Route. 
+  
+  Therefore, such a load balancing algorithm is out of scope for this enhancement and should be handled with a separate RFE and thereby it will be considered as a non-goal.
+
+#### Will this enhancement cover appending headers?
+  Answer: Yes and no.  This enhancement will not cover appending of
+  HTTP response and request headers as per the requirements from
+  [RFE-464](https://issues.redhat.com/browse/RFE-464). 
+  While this RFE mentions use-cases that require the ability to set or delete headers, there is no use-case
+  from this RFE that requires the ability to append headers. 
+
+  Once such a use-case arises, we may add the option to append headers, with due
+  consideration of any additional requirements deriving from the use-case
+  itself or considerations around the handling of comma-separated values.  (Note
+  that a use-case involving the `x-forwarded-*` headers, which is not mentioned
+  on the RFE, is described below.)
```

Comment:
Where is this use-case mentioned?

Reply:
just added in For Example in the next lines

Reply:
Is this what you mean by a use-case?
```markdown
  For the headers like `X-Forwarded-For` which might need values to be appended. 
  For Example:  X-Forwarded-For: <client>, <proxy1>, <proxy2>
  We have a different API for that.
```
It's difficult to understand what you are trying to say in this sub-section, or how it is different from the next sub-section ("For headers such as `X-Forwarded-For`, it makes sense to add an "Append" action. So why does this enhancement not cover appending?").  Maybe you can simplify:
```markdown
#### Will this enhancement cover appending headers?

Answer: No, this enhancement does not include an append action.
In order to keep its scope focused, the enhancement only adds set and delete actions.
These actions are sufficient to satisfy the use-cases from [RFE-464](https://issues.redhat.com/browse/RFE-464).
However, the API is designed to be extensible so that an append action, or other actions, can be added in the future, should the need arise.
```